### ISSUE TITLE
Db Migrate() throws Access denied for user 'root'@'localhost'

### ISSUE THREAD

**Issue by waseemdev** [PomeloFoundation/Pomelo.EntityFrameworkCore.MySql#908]
This exception occurs only when calling `Migrate/MigrateAsync`, reading data from tables, for example, is working without problems, here is my test that I used:
```csharp
    var count = await dbContext.TestTable.CountAsync(); // no problem
    Log(count);
    try
    {
        await dbContext.Database.MigrateAsync(); // exception!!!
    }
    catch (Exception ex)
    {
        LogError(ex);
    }
    count = await dbContext.TestTable.CountAsync(); // no problem
    Log(count);
```
Exception:
```
fail: Microsoft.EntityFrameworkCore.Database.Connection[20004]
      An error occurred using the connection to database '' on server 'localhost'.
MySql.Data.MySqlClient.MySqlException (0x80004005): Access denied for user 'root'@'localhost'
   at MySqlConnector.Core.ServerSession.TryAsyncContinuation(Task`1 task) in C:\projects\mysqlconnector\src\MySqlConnector\Core\ServerSession.cs:line 1245
   at System.Threading.Tasks.ContinuationResultTaskFromResultTask`2.InnerInvoke()
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location where exception was thrown ---
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot)
--- End of stack trace from previous location where exception was thrown ---
   at MySqlConnector.Core.ServerSession.ConnectAsync(ConnectionSettings cs, ILoadBalancer loadBalancer, IOBehavior ioBehavior, CancellationToken cancellationToken) in C:\projects\mysqlconnector\src\MySqlConnector\Core\ServerSession.cs:line 356
   at MySql.Data.MySqlClient.MySqlConnection.CreateSessionAsync(Nullable`1 ioBehavior, CancellationToken cancellationToken) in C:\projects\mysqlconnector\src\MySqlConnector\MySql.Data.MySqlClient\MySqlConnection.cs:line 416
   at MySql.Data.MySqlClient.MySqlConnection.OpenAsync(Nullable`1 ioBehavior, CancellationToken cancellationToken) in C:\projects\mysqlconnector\src\MySqlConnector\MySql.Data.MySqlClient\MySqlConnection.cs:line 175
   at Microsoft.EntityFrameworkCore.Storage.RelationalConnection.OpenDbConnectionAsync(Boolean errorsExpected, CancellationToken cancellationToken)
```

### Further technical details

MariaDB version: 10.4.8
Operating system: Ubuntu 18.04
Pomelo.EntityFrameworkCore.MySql version: 2.1.2
Microsoft.AspNetCore.App version: 2.1.2 (inside docker)

**Comment by lauxjpn** [comment]
I **could not reproduce** this issue with the following code in `2.1.2` (after creating an initial migration using the command line) or any newer version:

```c#
using System.Diagnostics;
using System.Linq;
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Logging;

namespace IssueConsoleTemplate
{
    public class TestEntity
    {
        public int Id { get; set; }
        public int MyValue { get; set; }
    }

    public class Issue908Context : DbContext
    {
        public DbSet<TestEntity> TestEntities { get; set; }

        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            // Use non-root non-default user with password to test.
            // The (empty) database "Issue908" must already exist and "issue908User" must have access.
            optionsBuilder
                .UseMySql("server=127.0.0.1;port=3308;user=issue908User;password=asdf;database=Issue908")
                .UseLoggerFactory(LoggerFactory.Create(b => b
                    .AddConsole()
                    .AddFilter(level => true)))
                .EnableSensitiveDataLogging();
        }

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            modelBuilder.Entity<TestEntity>()
                .HasData(new TestEntity { Id = 1, MyValue = 42 });
        }
    }

    internal class Program
    {
        private static void Main()
        {
            using var context = new Issue908Context();

            context.Database.Migrate();

            var theEntity = context.TestEntities.Single();

            Debug.Assert(theEntity.MyValue == 42);
        }
    }
}
```

Please provide us with a fully functional example, that reproduces this issue.

**Comment by waseemdev** [comment]
I spend the whole day figuring out what is the problem with no success.
Here is an example (https://github.com/waseemdev/mariadb-test) that doesn't work on the Ubuntu server but works on my local machine (Windows 10)!

Thank you.

**Comment by lauxjpn** [comment]
I will take a look at it.

Did you run the sample code I provided? Does it work or does it fail to run?
Did you try run your project outside of your docker container on Ubuntu (e.g. in a local VM)?

**Comment by lauxjpn** [comment]
Your issue is related to your shared `DbContext` code.
I made the following changes to your files to make your test methods work:

**Startup.cs:**
```c#
using System;
using Microsoft.AspNetCore.Builder;
using Microsoft.AspNetCore.Hosting;
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;

namespace WebApplication6
{
    public class Startup
    {
        public Startup(IConfiguration configuration)
        {
            Configuration = configuration;
        }

        public IConfiguration Configuration { get; }

        // This method gets called by the runtime. Use this method to add services to the container.
        public void ConfigureServices(IServiceCollection services)
        {
            var connectionString = Configuration["Data:ApplicationDbContext:ConnectionString"];

            services.AddEntityFrameworkMySql()
                .AddDbContext<ApplicationDbContext>(options => options
                    .UseMySql(connectionString, sqlOptions => {
                        sqlOptions.MigrationsAssembly("WebApplication6");
                        sqlOptions.EnableRetryOnFailure(
                            maxRetryCount: 4,
                            maxRetryDelay: TimeSpan.FromMilliseconds(2000),
                            errorNumbersToAdd: null);
                    })
                    .EnableSensitiveDataLogging()
                    .EnableDetailedErrors());

            services.AddMvc();
        }

        // This method gets called by the runtime. Use this method to configure the HTTP request pipeline.
        public void Configure(IApplicationBuilder app, IHostingEnvironment env)
        {
            if (env.IsDevelopment())
            {
                //app.UseBrowserLink();
                app.UseDeveloperExceptionPage();
            }
            else
            {
                app.UseExceptionHandler("/Error");
            }

            app.UseStaticFiles();

            app.UseMvc();
        }
    }
}
```

**ApplicationDbContext.cs:**
```c#
using Microsoft.EntityFrameworkCore;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace WebApplication6
{
    public class ApplicationDbContext : DbContext
    {
        public ApplicationDbContext(DbContextOptions<ApplicationDbContext> options)
            : base(options)
        {
        }

        public DbSet<Items> Items { get; set; }
    }

    public class Items
    {
        [Key]
        [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
        public long Id { get; set; }
        public string Name { get; set; }
        public string Name2 { get; set; }
        public string Name3 { get; set; }
    }
}
```

**Comment by waseemdev** [comment]
> Did you run the sample code I provided? Does it work or does it fail?

No, but I tried my example with normal `DbContext` setup without the shared `DbContext` code as you mentioned, and it is working.
I added this code to share the `MySqlConnection` between `DbContext`s so I can use the same transaction over multiple `DbContext`s.

> Did you try to run your project outside of your docker container on Ubuntu (e.g. in a local VM)?

No.

**Comment by lauxjpn** [comment]
We improved the transaction support in `3.0.0`, including the addition of ambient transactions. So if you use the latest prerelease, the following instructions from the official docs should work:
https://docs.microsoft.com/en-us/ef/core/saving/transactions#cross-context-transaction-relational-databases-only

**Comment by waseemdev** [comment]
Ok, since the issue is when using migration with MySqlConnection only and not with ConnectionString, and at the migration time no need for the transaction, I came up with a simple workaround by adding IsMigration property inside DbContext and checking it, if true then will use a connection string, otherwise will use the shared connection:
```csharp
    if (IsMigration)
        optionsBuilder.UseMySql(connectionString, ...);
    else
        optionsBuilder.UseMySql(serviceProvider.GetRequiredService<MySqlConnection>(), ...);
```

Thank you for the information.

**Comment by roji** [comment]
@lauxjpn yeah, this is indeed an issue. In Npgsql I worked around this by having a CloneWith method on NpgsqlConnection, which accepts a new connection string but keeps authentication information from the clone source (password, SSL/TLS callbacks if relevant). This also allows you to take an existing connection string and change options on it (e.g. make it non-pooling) while conserving auth options, regardless of whether PersistSecurityInfo was set to false.

/cc @bgrainger

**Comment by bgrainger** [comment]
> But if you called it with an already existing connection object, it will retrieve the connection string used for this original connection object from the object itself at the point in time when the temporary connection object needs to be created.

`MySqlConnection.Clone` was implemented in 0.60.0: https://github.com/mysql-net/MySqlConnector/issues/720

This should copy the original connection string, including authentication information. If you have code that looks like:

```
var newConnection = new MySqlConnection(existingConnection.ConnectionString);
await newConnection.OpenAsync();
```

with 0.60.0, you may be able to replace it with:

```
var newConnection = existingConnection.Clone();
await newConnection.OpenAsync();
```

Would this fix the problem here?

**Comment by lauxjpn** [comment]
@bgrainger Yes, that will work!

**Comment by bgrainger** [comment]
FWIW, I just realised there's a potential problem with `Clone` (related to getting the connection password): https://github.com/mysql-net/MySqlConnector/issues/735

I don't think it's serious, since any code that can call `Clone` has other ways (e.g., reflection) to get at that data. But you may want to wait until 0.60.4 is released (later today) if it's a concern.

**Comment by lauxjpn** [comment]
I think https://github.com/mysql-net/MySqlConnector/issues/735 would remove the possibility again for us to change the connection string settings for a cloned connection and still keep sensitive information, am I right?
The implementation I would have used for the pre 735 fix would have been the following:

```c#
public virtual IMySqlRelationalConnection CreateMasterConnection()
{
    var relationalOptions = RelationalOptionsExtension.Extract(Dependencies.ContextOptions);

    MySqlConnection connection;
    string connectionString;

    // Make sure to get the original connection string including security sensitive information.
    if (relationalOptions.Connection == null)
    {
        connection = null;
        connectionString = ConnectionString;
    }
    else
    {
        connection = ((MySqlConnection)DbConnection).Clone(); // <-- Clone the connection
        connectionString = connection.ConnectionString; // <-- Connection string with password
    }

    // Add master connection specific options.
    var csb = new MySqlConnectionStringBuilder(connectionString) {
        Database = string.Empty,
        Pooling = false
    };

    // Apply modified connection string.
    if (connection == null)
    {
        relationalOptions = relationalOptions.WithConnectionString(csb.ConnectionString);
    }
    else
    {
        connection.ConnectionString = csb.ConnectionString;
        relationalOptions = relationalOptions.WithConnection(connection);
    }

    var optionsBuilder = new DbContextOptionsBuilder();
    ((IDbContextOptionsBuilderInfrastructure)optionsBuilder).AddOrUpdateExtension(relationalOptions);

    return new MySqlRelationalConnection(Dependencies.With(optionsBuilder.Options), _serviceProvider)
    {
        IsMasterConnection = true
    };
}
```
I agree though, that MySqlConnector's current implementation does allow to directly retrieve the password, which would violate the whole idea of stripping it from the string in the first place.

If MySqlConnector would implement a method (or overload) like @roji's `CloneWith(string connectionString)`, that takes a connection string as an argument and merges it with its internal one, we would be able supply different options without needing the password retrieved.

https://github.com/npgsql/efcore.pg/blob/55bc10d0c131afea18fd9e820d5924c464910afb/src/EFCore.PG/Storage/Internal/NpgsqlRelationalConnection.cs#L60

```c#
relationalOptions.WithConnection(((NpgsqlConnection)DbConnection).CloneWith(connectionString))
```

**Comment by bgrainger** [comment]
Yes, MySqlConnector 0.60.4 will remove the loophole that the code above is (planning on) relying on.

Adding `CloneWith` seems like the best way to solve this problem. Opened https://github.com/mysql-net/MySqlConnector/issues/736.

**Comment by bgrainger** [comment]
`CloneWith` is now available in 0.61.0. I think the code to use it would be a bit shorter and look like this:

```csharp
    // Get current connection string.
    string connectionString = relationalOptions.Connection?.ConnectionString ?? ConnectionString;

    // Add master connection specific options.
    var csb = new MySqlConnectionStringBuilder(connectionString) {
        Database = "",
        Pooling = false,
    };

    // Apply modified connection string.
    if (relationalOptions.Connection is null)
    {
        relationalOptions = relationalOptions.WithConnectionString(csb.ConnectionString);
    }
    else
    {
        var connection = relationalOptions.Connection.CloneWith(csb.ConnectionString);
        relationalOptions = relationalOptions.WithConnection(connection);
    }
```

**Comment by lauxjpn** [comment]
@waseemdev Thanks you for taking the time to report this issue!
And special thanks to @bgrainger and @roji for your work and input on this issue and its solution.

This is fixed with #916.
